cunumeric/array.py

```diff
@@ -3113,32 +3138,38 @@ def mean(
                 dtype = np.dtype(np.float64)
             else:
                 dtype = self.dtype
+
+        where_array = broadcast_where(where, self.shape)
         # Do the sum
         if out is not None and out.dtype == dtype:
             sum_array = self.sum(
-                axis=axis,
-                dtype=dtype,
-                out=out,
-                keepdims=keepdims,
+                axis=axis, out=out, keepdims=keepdims, where=where_array
             )
         else:
             sum_array = self.sum(
-                axis=axis,
-                dtype=dtype,
-                keepdims=keepdims,
+                axis=axis, keepdims=keepdims, where=where_array
             )
+
         if axis is None:
-            divisor = reduce(lambda x, y: x * y, self.shape, 1)
+            if where_array is not None:
+                divisor = where_array._count_nonzero()
+            else:
+                divisor = reduce(lambda x, y: x * y, self.shape, 1)
+
         else:
-            divisor = self.shape[axis]
+            if where_array is not None:
+                divisor = where_array._count_nonzero(axis=axis)
+            else:
+                divisor = self.shape[axis]
         # Divide by the number of things in the collapsed dimensions
         # Pick the right kinds of division based on the dtype
+        sum_array = sum_array.astype(dtype)
         if dtype.kind == "f" or dtype.kind == "c":
             sum_array.__itruediv__(
-                np.array(divisor, dtype=sum_array.dtype),
+                np.array(divisor, dtype=dtype),
```

Comment:
It is no longer a good idea to convert to a NumPy array indiscriminately, since `divisor` can now be a cuNumeric array. Suggest doing it as follows here and in `__ifloordiv__` below, and making `divisor` an array directly above.
```suggestion
                divisor.astype(dtype),
```